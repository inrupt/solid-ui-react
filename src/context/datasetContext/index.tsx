//
// Copyright Inrupt Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to use,
// copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
// Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

import React, { createContext, ReactElement, useState, useEffect } from "react";
import {
  SolidDataset,
  WithResourceInfo,
  UrlString,
} from "@inrupt/solid-client";

import useDataset from "../../hooks/useDataset";

export interface IDatasetContext {
  solidDataset: SolidDataset | (SolidDataset & WithResourceInfo) | undefined;
  setDataset: (solidDataset: SolidDataset) => void;
}

const DatasetContext = createContext<IDatasetContext>({
  solidDataset: undefined,
  setDataset: () => {},
});

export default DatasetContext;

export interface IDatasetProvider {
  children: React.ReactNode;
  /**
   * @deprecated `loading` is deprecated. It can still be used but may be removed in a future major release.`loadingComponent` should be used instead.
   */
  loading?: React.ReactNode;
  loadingComponent?: React.ComponentType | null;
  onError?(error: Error): void | null;
  solidDataset?: SolidDataset | (SolidDataset & WithResourceInfo);
  datasetUrl?: UrlString | string;
}

export type RequireProperty<T, Prop extends keyof T> = T & {
  [key in Prop]-?: T[key];
};

export type RequireDatasetOrDatasetUrl =
  | RequireProperty<IDatasetProvider, "solidDataset">
  | RequireProperty<IDatasetProvider, "datasetUrl">;

/**
 * Used to provide a [Dataset](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-SolidDataset) to child components through context, as used by various provided components and the useDataset hook.
 */
export const DatasetProvider = ({
  children,
  onError,
  solidDataset: propDataset,
  datasetUrl,
  loading,
  loadingComponent: LoadingComponent,
}: RequireDatasetOrDatasetUrl): ReactElement => {
  const { dataset, error } = useDataset(datasetUrl);

  if (error && onError) {
    onError(error);
  }

  const datasetToUse = propDataset ?? dataset;

  // Provide a setDataset function so that child components can update.
  const [stateDataset, setDataset] = useState<
    SolidDataset | (SolidDataset & WithResourceInfo) | undefined
  >(datasetToUse);

  // If the dataset is asynchronously loaded, make sure to set the new state value.
  useEffect(() => {
    setDataset(datasetToUse);
  }, [datasetToUse]);

  let loader: JSX.Element | React.ReactNode | null = (LoadingComponent && (
    <LoadingComponent />
  )) ||
    loading || <span>Fetching data...</span>;

  if (LoadingComponent === null) {
    loader = null;
  }

  return (
    <DatasetContext.Provider value={{ solidDataset: stateDataset, setDataset }}>
      {stateDataset ? children : loader}
    </DatasetContext.Provider>
  );
};
