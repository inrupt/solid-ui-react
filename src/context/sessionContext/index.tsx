/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React, {
  createContext,
  ReactElement,
  useState,
  SetStateAction,
  Dispatch,
  useEffect,
  ReactNode,
} from "react";

import {
  fetch,
  login,
  logout,
  handleIncomingRedirect,
  Session,
  getDefaultSession,
  onSessionRestore as onSessionRestoreClient,
} from "@inrupt/solid-client-authn-browser";

import { ILoginInputOptions } from "@inrupt/solid-client-authn-core";

export interface ISessionContext {
  login: (options: ILoginInputOptions) => Promise<void>;
  logout: () => Promise<void>;
  session: Session;
  sessionRequestInProgress: boolean;
  setSessionRequestInProgress?: Dispatch<SetStateAction<boolean>> | any;
  fetch: typeof window.fetch;
}

export const SessionContext = createContext<ISessionContext>({
  login,
  logout,
  fetch,
  session: getDefaultSession(),
  sessionRequestInProgress: true,
});

/* eslint react/require-default-props: 0 */
export interface ISessionProvider {
  children: ReactNode;
  sessionId?: string;
  session?: Session;
  sessionRequestInProgress?: boolean;
  onError?: (error: Error) => void;
  restorePreviousSession?: boolean;
  onSessionRestore?: (url: string) => void;
}

/**
 * Used to provide session data to child components through context, as used by various provided components and the useSession hook.
 */
export const SessionProvider = ({
  sessionId,
  children,
  onError,
  sessionRequestInProgress: defaultSessionRequestInProgress,
  restorePreviousSession,
  onSessionRestore,
}: ISessionProvider): ReactElement => {
  const [session, setSession] = useState<Session>(getDefaultSession());

  if (onSessionRestore !== undefined) {
    onSessionRestoreClient(onSessionRestore);
  }

  const defaultInProgress =
    typeof defaultSessionRequestInProgress === "undefined"
      ? !session.info.isLoggedIn
      : defaultSessionRequestInProgress;

  // If loggedin is true, we're not making a session request.
  const [sessionRequestInProgress, setSessionRequestInProgress] = useState(
    defaultInProgress
  );

  let currentLocation;

  if (typeof window !== "undefined") {
    currentLocation = window.location;
  }

  useEffect(() => {
    handleIncomingRedirect({
      url: window.location.href,
      restorePreviousSession,
    })
      .catch((error: Error) => {
        if (onError) {
          onError(error);
        } else {
          throw error;
        }
      })
      .finally(() => {
        // console.log("done");
        setSessionRequestInProgress(false);
      });

    getDefaultSession().on("logout", () => {
      // TODO force a refresh
      setSession(getDefaultSession());
    });
  }, [session, sessionId, onError, currentLocation, restorePreviousSession]);

  const contextLogin = async (options: ILoginInputOptions) => {
    setSessionRequestInProgress(true);

    try {
      await login(options);
    } catch (error) {
      if (onError) {
        onError(error);
      } else {
        throw error;
      }
    } finally {
      setSessionRequestInProgress(false);
    }
  };

  const contextLogout = async () => {
    try {
      await logout();
    } catch (error) {
      if (onError) {
        onError(error);
      } else {
        throw error;
      }
    }
  };

  return (
    <SessionContext.Provider
      value={{
        session,
        login: contextLogin,
        logout: contextLogout,
        sessionRequestInProgress,
        setSessionRequestInProgress,
        fetch,
      }}
    >
      {children}
    </SessionContext.Provider>
  );
};
