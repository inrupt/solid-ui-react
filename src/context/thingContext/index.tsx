/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React, {
  createContext,
  ReactElement,
  useState,
  useEffect,
  useContext,
} from "react";
import {
  Thing,
  LitDataset,
  WithResourceInfo,
  UrlString,
  getThingOne,
} from "@inrupt/solid-client";
import { DatasetContext } from "../datasetContext";

interface IThingContext {
  thing: Thing | undefined;
}

export const ThingContext = createContext<IThingContext>({
  thing: undefined,
});

interface IThingProvider {
  children: React.ReactNode;
  thing?: Thing;
  thingUrl?: UrlString | string;
  dataset?: LitDataset | (LitDataset & WithResourceInfo);
  datasetUrl?: UrlString | string;
}

type RequireProperty<T, Prop extends keyof T> = T & { [key in Prop]-?: T[key] };
type RequireDatasetOrDatasetUrl =
  | RequireProperty<IThingProvider, "thing">
  | RequireProperty<IThingProvider, "thingUrl">;

export const ThingProvider = ({
  children,
  thing,
  thingUrl,
}: RequireDatasetOrDatasetUrl): ReactElement => {
  const datasetContext = useContext(DatasetContext);
  const { dataset } = datasetContext;
  const [litThing, setLitThing] = useState<Thing>();

  useEffect(() => {
    if (dataset && thingUrl) {
      setLitThing(getThingOne(dataset, thingUrl));
    }
  }, [dataset, thingUrl]);

  if (thing) {
    return (
      <ThingContext.Provider
        value={{
          thing,
        }}
      >
        {children}
      </ThingContext.Provider>
    );
  }

  return (
    <ThingContext.Provider
      value={{
        thing: litThing,
      }}
    >
      {children}
    </ThingContext.Provider>
  );
};
