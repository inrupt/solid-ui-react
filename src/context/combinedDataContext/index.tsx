//
// Copyright Inrupt Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to use,
// copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
// Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

import type { ReactElement } from "react";
import React from "react";

import type {
  Thing,
  SolidDataset,
  WithResourceInfo,
  UrlString,
} from "@inrupt/solid-client";

import { DatasetProvider } from "../datasetContext";
import { ThingProvider } from "../thingContext";

export type Props = {
  children: React.ReactNode;
  loadingComponent?: React.ComponentType | null;
  onError?(error: Error): void | null;
};

export type ThingOrThingUrl = {
  thing?: Thing;
  thingUrl?: UrlString | string;
};

export interface Dataset extends Props {
  solidDataset: SolidDataset | (SolidDataset & WithResourceInfo);
}

export interface DatasetUrl extends Props {
  datasetUrl: UrlString | string;
}

export type RequireProperty<T, Prop extends keyof T> = T & {
  [key in Prop]-?: T[key];
};

export type RequireThingOrThingUrl =
  | RequireProperty<ThingOrThingUrl, "thing">
  | RequireProperty<ThingOrThingUrl, "thingUrl">;

// TODO: naming of this provider TBC
function CombinedDataProvider(
  props: Dataset & RequireThingOrThingUrl,
): ReactElement;

function CombinedDataProvider(
  props: DatasetUrl & RequireThingOrThingUrl,
): ReactElement;

/**
 * Used to provide both a [Dataset](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-SolidDataset) and [Thing](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-Thing) to child components through context, as used by various provided components and the useDataset and useThing hooks.
 */
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function CombinedDataProvider(props: any): ReactElement {
  const {
    children,
    solidDataset,
    datasetUrl,
    thing,
    thingUrl,
    onError,
    loadingComponent,
  } = props;

  return (
    <DatasetProvider
      onError={onError}
      solidDataset={solidDataset}
      datasetUrl={datasetUrl}
      loadingComponent={loadingComponent}
    >
      <ThingProvider thing={thing} thingUrl={thingUrl}>
        {children}
      </ThingProvider>
    </DatasetProvider>
  );
}

export default CombinedDataProvider;
