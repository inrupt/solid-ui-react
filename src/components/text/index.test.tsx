//
// Copyright Inrupt Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to use,
// copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
// Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

import { jest, it, describe, expect } from "@jest/globals";
import * as React from "react";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { ErrorBoundary } from "react-error-boundary";
import SolidFns from "@inrupt/solid-client";
import type * as SolidClient from "@inrupt/solid-client";
import * as helpers from "../../helpers";
import { Text } from ".";
import { DatasetProvider } from "../../context/datasetContext";
import { ThingProvider } from "../../context/thingContext";

const mockDatasetFromWithChangelog = (url: string) =>
  SolidFns.mockSolidDatasetFrom(url) as SolidClient.SolidDataset &
    SolidClient.WithServerResourceInfo &
    // Pretend we have a changelog, even though we don't, as it's irrelevant to these tests.
    SolidClient.WithChangeLog;

jest.mock("@inrupt/solid-client", () => {
  const solidClient = jest.requireActual(
    "@inrupt/solid-client",
  ) as typeof SolidClient;
  return {
    ...solidClient,
    // Set a default mocked value for saveSolidDatasset, that may be overriden at the test scope.
    saveSolidDatasetAt: jest
      .fn<(typeof SolidClient)["saveSolidDatasetAt"]>()
      .mockResolvedValue(
        solidClient.mockSolidDatasetFrom(
          "https://some-interesting-value.com",
        ) as Awaited<ReturnType<typeof solidClient.saveSolidDatasetAt>>,
      ),
    getSolidDataset: jest
      .fn<(typeof SolidClient)["getSolidDataset"]>()
      .mockResolvedValue(
        solidClient.mockSolidDatasetFrom(
          "https://some-interesting-value.com",
        ) as Awaited<ReturnType<typeof solidClient.getSolidDataset>>,
      ),
  };
});

const mockPredicate = `http://xmlns.com/foaf/0.1/nick`;
const mockNick = "test nick value";

const defaultMockedThing = () =>
  SolidFns.addStringNoLocale(SolidFns.createThing(), mockPredicate, mockNick);

const defaultMockedDataset = () =>
  SolidFns.setThing(
    SolidFns.mockSolidDatasetFrom("https://some-interesting-value.com"),
    defaultMockedThing(),
  );

const defaultmockedDatasetWithResourceInfo = () =>
  SolidFns.setThing(
    SolidFns.mockSolidDatasetFrom("https://some-interesting-value.com"),
    defaultMockedThing(),
  );

const inputOptions = {
  name: "test-name",
  type: "url",
};

describe("<Text /> component snapshot test", () => {
  it("matches snapshot", () => {
    const documentBody = render(
      <Text
        solidDataset={defaultmockedDatasetWithResourceInfo()}
        thing={defaultMockedThing()}
        property={mockPredicate}
      />,
    );
    const { baseElement } = documentBody;
    expect(baseElement).toMatchSnapshot();
  });

  it("matches snapshot with data from context", () => {
    const documentBody = render(
      <DatasetProvider solidDataset={defaultmockedDatasetWithResourceInfo()}>
        <ThingProvider thing={defaultMockedThing()}>
          <Text property={mockPredicate} edit autosave />
        </ThingProvider>
      </DatasetProvider>,
    );
    const { baseElement } = documentBody;
    expect(baseElement).toMatchSnapshot();
  });

  it("matches snapshot with edit true and inputOptions", () => {
    const documentBody = render(
      <Text
        edit
        inputProps={inputOptions}
        solidDataset={defaultmockedDatasetWithResourceInfo()}
        thing={defaultMockedThing()}
        property={mockPredicate}
      />,
    );
    const { baseElement } = documentBody;
    expect(baseElement).toMatchSnapshot();
  });
  it("matches snapshot while fetching data", () => {
    jest.spyOn(helpers, "useProperty").mockReturnValueOnce({
      thing: undefined,
      error: undefined,
      value: null,
      setDataset: jest.fn(),
      setThing: jest.fn(),
      property: mockPredicate,
    });
    const documentBody = render(<Text property={mockPredicate} />);
    const { baseElement } = documentBody;
    expect(baseElement).toMatchSnapshot();
  });

  it("renders loading component if passed while fetching data", () => {
    jest.spyOn(helpers, "useProperty").mockReturnValueOnce({
      thing: undefined,
      error: undefined,
      value: null,
      setDataset: jest.fn(),
      setThing: jest.fn(),
      property: mockPredicate,
    });
    const documentBody = render(
      <Text
        property={mockPredicate}
        loadingComponent={() => (
          <span id="custom-loading-component">loading...</span>
        )}
      />,
    );
    const { baseElement } = documentBody;
    expect(baseElement).toMatchSnapshot();
  });

  it("does not render default loading message if loadingComponent is null", () => {
    jest.spyOn(helpers, "useProperty").mockReturnValueOnce({
      thing: undefined,
      error: undefined,
      value: null,
      setDataset: jest.fn(),
      setThing: jest.fn(),
      property: mockPredicate,
    });
    const documentBody = render(
      <Text property={mockPredicate} loadingComponent={null} />,
    );
    const { baseElement } = documentBody;
    expect(baseElement).toMatchSnapshot();
  });

  it("renders default error message if there is an error", () => {
    const emptyThing = SolidFns.createThing();
    const documentBody = render(
      <Text thing={emptyThing} property="https://example.com/bad-url" />,
    );
    const { baseElement } = documentBody;
    expect(baseElement).toMatchSnapshot();
  });

  it("renders custom error component if passed and there is an error", () => {
    const emptyThing = SolidFns.createThing();
    const documentBody = render(
      <Text
        thing={emptyThing}
        property="https://example.com/bad-url"
        errorComponent={({ error }) => (
          <span id="custom-error-component">{error.toString()}</span>
        )}
      />,
    );
    const { baseElement } = documentBody;
    expect(baseElement).toMatchSnapshot();
  });
});

describe("<Text /> component functional testing", () => {
  it("Should call getStringNoLocale function if no locale is passed", async () => {
    const mockThing = SolidFns.addStringNoLocale(
      SolidFns.createThing(),
      mockPredicate,
      mockNick,
    );

    const mockDataset = SolidFns.setThing(
      SolidFns.mockSolidDatasetFrom("https://some-interesting-value.com"),
      mockThing,
    );

    jest.spyOn(SolidFns, "getStringNoLocale");

    const { getByText } = render(
      <Text
        solidDataset={mockDataset}
        thing={mockThing}
        property={mockPredicate}
      />,
    );
    expect(SolidFns.getStringNoLocale).toHaveBeenCalled();
    expect(getByText(mockNick)).toBeDefined();
  });

  it("Should call getStringWithLocale function if locale is passed", async () => {
    const mockThing = SolidFns.addStringWithLocale(
      SolidFns.createThing(),
      mockPredicate,
      mockNick,
      "en",
    );

    const mockDataset = SolidFns.setThing(
      SolidFns.mockSolidDatasetFrom("https://some-interesting-value.com"),
      mockThing,
    );

    jest.spyOn(SolidFns, "getStringWithLocale");

    const { getByText } = render(
      <Text
        solidDataset={mockDataset}
        thing={mockThing}
        property={mockPredicate}
        locale="en"
        autosave
      />,
    );
    expect(SolidFns.getStringWithLocale).toHaveBeenCalled();
    expect(getByText(mockNick)).toBeDefined();
  });

  it("Should call setStringNoLocale onBlur if no locale is set", async () => {
    const mockThing = SolidFns.addStringNoLocale(
      SolidFns.createThing(),
      mockPredicate,
      mockNick,
    );

    const mockDataset = SolidFns.setThing(
      SolidFns.mockSolidDatasetFrom("https://some-interesting-value.com"),
      mockThing,
    );

    jest.spyOn(SolidFns, "setStringNoLocale");

    const user = userEvent.setup();
    const { getByDisplayValue } = render(
      <Text
        solidDataset={mockDataset}
        thing={mockThing}
        property={mockPredicate}
        edit
        autosave
      />,
    );
    const input = getByDisplayValue(mockNick);
    input.focus();

    await user.type(input, "updated nick value");
    // blur the input:
    await user.tab();

    expect(SolidFns.setStringNoLocale).toHaveBeenCalled();
  });

  it("Should call setStringWithLocale onBlur if locale is set", async () => {
    const mockThing = SolidFns.addStringWithLocale(
      SolidFns.createThing(),
      mockPredicate,
      mockNick,
      "en",
    );

    const mockDataset = SolidFns.setThing(
      SolidFns.mockSolidDatasetFrom("https://some-interesting-value.com"),
      mockThing,
    );

    jest.spyOn(SolidFns, "setStringWithLocale");

    const user = userEvent.setup();
    const { getByDisplayValue } = render(
      <Text
        solidDataset={mockDataset}
        thing={mockThing}
        property={mockPredicate}
        locale="en"
        edit
        autosave
      />,
    );
    const input = getByDisplayValue(mockNick);
    input.focus();

    await user.type(input, "updated nick value");
    // blur the input:
    await user.tab();

    expect(SolidFns.setStringWithLocale).toHaveBeenCalled();
  });

  it("Should not call setString onBlur if the value of the input hasn't changed", async () => {
    const mockThing = SolidFns.addStringWithLocale(
      SolidFns.createThing(),
      mockPredicate,
      mockNick,
      "en",
    );

    const mockDataset = SolidFns.setThing(
      SolidFns.mockSolidDatasetFrom("https://some-interesting-value.com"),
      mockThing,
    );

    jest.spyOn(SolidFns, "setStringWithLocale");

    const user = userEvent.setup();
    const { getByDisplayValue } = render(
      <Text
        solidDataset={mockDataset}
        thing={mockThing}
        property={mockPredicate}
        locale="en"
        edit
        autosave
      />,
    );
    const input = getByDisplayValue(mockNick);
    input.focus();

    await user.type(input, "u{Backspace}");
    // blur the input:
    await user.tab();

    expect(SolidFns.setStringWithLocale).not.toHaveBeenCalled();
  });

  it("Should not call saveSolidDatasetAt onBlur if autosave is false", async () => {
    const mockThing = SolidFns.addStringWithLocale(
      SolidFns.createThing(),
      mockPredicate,
      mockNick,
      "en",
    );

    const mockDataset = SolidFns.setThing(
      SolidFns.mockSolidDatasetFrom("https://some-interesting-value.com"),
      mockThing,
    );

    jest.spyOn(SolidFns, "saveSolidDatasetAt");
    const { getByDisplayValue } = render(
      <Text
        solidDataset={mockDataset}
        thing={mockThing}
        property={mockPredicate}
        locale="en"
        edit
      />,
    );
    getByDisplayValue(mockNick).focus();
    getByDisplayValue(mockNick).blur();
    expect(SolidFns.saveSolidDatasetAt).toHaveBeenCalledTimes(0);
  });

  it("Should call onSave if it is passed", async () => {
    const onSave = jest.fn();
    const onError = jest.fn();

    const user = userEvent.setup();

    const { getByDisplayValue } = render(
      <Text
        solidDataset={defaultmockedDatasetWithResourceInfo()}
        thing={defaultMockedThing()}
        property={mockPredicate}
        onSave={onSave}
        onError={onError}
        edit
        autosave
      />,
    );
    const input = getByDisplayValue(mockNick);
    input.focus();

    await user.type(input, "updated nick value");
    // blur the input:
    await user.tab();

    expect(onSave).toHaveBeenCalled();
  });

  it("Should call onSave for fetched dataset with custom location if it is passed", async () => {
    const onSave = jest.fn();
    const onError = jest.fn();

    const user = userEvent.setup();
    const { getByDisplayValue } = render(
      <Text
        solidDataset={defaultmockedDatasetWithResourceInfo()}
        thing={defaultMockedThing()}
        property={mockPredicate}
        saveDatasetTo="https://ldp.demo-ess.inrupt.com/norbertand/profile/card"
        onSave={onSave}
        onError={onError}
        edit
        autosave
      />,
    );
    const input = getByDisplayValue(mockNick);
    input.focus();

    await user.type(input, "updated nick value");

    // blur the input:
    await user.tab();

    expect(onSave).toHaveBeenCalled();
  });

  it("Should call onError if saving fails", async () => {
    const { saveSolidDatasetAt: mockedSaveDataset } = jest.requireMock(
      "@inrupt/solid-client",
    ) as jest.Mocked<typeof SolidClient>;
    mockedSaveDataset.mockRejectedValueOnce(null);
    const onError = jest.fn();

    const user = userEvent.setup();
    const { getByDisplayValue } = render(
      <Text
        solidDataset={defaultmockedDatasetWithResourceInfo()}
        thing={defaultMockedThing()}
        property={mockPredicate}
        saveDatasetTo="https://ldp.demo-ess.inrupt.com/norbertand/profile/card"
        onError={onError}
        edit
        autosave
      />,
    );
    const input = getByDisplayValue(mockNick);
    input.focus();

    await user.type(input, "updated nick value");

    // blur the input:
    await user.tab();

    expect(onError).toHaveBeenCalled();
  });

  it("Should call onError if saving fetched dataset to custom location fails", async () => {
    const { saveSolidDatasetAt: mockedSaveDataset } = jest.requireMock(
      "@inrupt/solid-client",
    ) as jest.Mocked<typeof SolidClient>;
    mockedSaveDataset.mockRejectedValueOnce(null);
    const onError = jest.fn();

    const user = userEvent.setup();
    const { getByDisplayValue } = render(
      <Text
        solidDataset={defaultmockedDatasetWithResourceInfo()}
        thing={defaultMockedThing()}
        property={mockPredicate}
        saveDatasetTo="https://ldp.demo-ess.inrupt.com/norbertand/profile/card"
        onError={onError}
        edit
        autosave
      />,
    );
    const input = getByDisplayValue(mockNick);
    input.focus();

    await user.type(input, "updated nick value");

    // blur the input:
    await user.tab();

    expect(onError).toHaveBeenCalled();
  });

  it("Should throw error if SaveDatasetTo is missing for new data", async () => {
    jest.spyOn(console, "error").mockImplementation(() => {});

    const user = userEvent.setup();
    const { getByText, getByDisplayValue } = render(
      <ErrorBoundary
        fallbackRender={({ error }) => <div>{JSON.stringify(error)}</div>}
      >
        <Text
          solidDataset={SolidFns.createSolidDataset()}
          thing={defaultMockedThing()}
          property={mockPredicate}
          edit
          autosave
        />
      </ErrorBoundary>,
    );

    const input = getByDisplayValue(mockNick);
    input.focus();

    await user.type(input, "updated nick value");

    // blur the input:
    await user.tab();

    expect(getByText("{}")).toBeDefined();
    // eslint-disable-next-line no-console
    (console.error as jest.Mock).mockRestore();
  });
  it("Should update context with latest dataset after saving", async () => {
    const { getSolidDataset: mockedGet } = jest.requireMock(
      "@inrupt/solid-client",
    ) as jest.Mocked<typeof SolidClient>;
    const mockedLatestDataset = mockDatasetFromWithChangelog(
      "https://some.irrelevant.url",
    );
    mockedGet.mockResolvedValueOnce(mockedLatestDataset);
    const setDataset = jest.fn();
    const setThing = jest.fn();
    jest.spyOn(helpers, "useProperty").mockReturnValue({
      dataset: defaultMockedDataset(),
      setDataset,
      setThing,
      error: undefined,
      value: mockNick,
      thing: defaultMockedThing(),
      property: mockPredicate,
    });

    const user = userEvent.setup();
    const { getByDisplayValue } = render(
      <Text
        solidDataset={defaultmockedDatasetWithResourceInfo()}
        thing={defaultMockedThing()}
        property={mockPredicate}
        locale="en"
        edit
        autosave
      />,
    );
    const input = getByDisplayValue(mockNick);
    input.focus();

    await user.type(input, "updated nick value");
    // blur the input:
    await user.tab();

    expect(SolidFns.saveSolidDatasetAt).toHaveBeenCalled();
    expect(setDataset).toHaveBeenCalledWith(mockedLatestDataset);
  });
});
