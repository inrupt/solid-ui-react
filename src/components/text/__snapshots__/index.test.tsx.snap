// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Text /> component snapshot test does not render defualt loading message if loadingComponent is null 1`] = `
<body>
  <div />
</body>
`;

exports[`<Text /> component snapshot test matches snapshot 1`] = `
<body>
  <div>
    <span>
      test nick value
    </span>
  </div>
</body>
`;

exports[`<Text /> component snapshot test matches snapshot while fetching data 1`] = `
<body>
  <div>
    <span>
      fetching data in progress
    </span>
  </div>
</body>
`;

exports[`<Text /> component snapshot test matches snapshot with data from context 1`] = `
<body>
  <div>
    <input
      type="text"
      value="test nick value"
    />
  </div>
</body>
`;

exports[`<Text /> component snapshot test matches snapshot with edit true and inputOptions 1`] = `
<body>
  <div>
    <input
      name="test-name"
      type="url"
      value="test nick value"
    />
  </div>
</body>
`;

exports[`<Text /> component snapshot test renders custom error component if passed and there is an error 1`] = `
<body>
  <div>
    <span
      id="custom-error-component"
    >
      Error: No value found for property.
    </span>
  </div>
</body>
`;

exports[`<Text /> component snapshot test renders default error message if there is an error 1`] = `
<body>
  <div>
    <span>
      Error: No value found for property.
    </span>
  </div>
</body>
`;

exports[`<Text /> component snapshot test renders loading component if passed while fetching data 1`] = `
<body>
  <div>
    <span
      id="custom-loading-component"
    >
      loading...
    </span>
  </div>
</body>
`;
